lib/rack/cors.rb

```diff
@@ -117,7 +117,7 @@ def call(env)
         else
           DEFAULT_VARY_HEADERS
         end
-        headers[VARY] = ((vary ? vary.split(/,\s*/) : []) + cors_vary_headers).uniq.join(', ')
+        headers = headers.merge(VARY => ((vary ? vary.split(/,\s*/) : []) + cors_vary_headers).uniq.join(', '))
```

Comment:
Using `.merge` creates a new hash, which will bloat memory. As of version 3, the rack specification states that headers must be an unfrozen hash, see https://github.com/rack/rack/blob/main/CHANGELOG.md#spec-changes-2

Thus, once you're on Rack 3, this issue should resolve itself, and we don't need to bloat memory "just in case".

Reply:
If you must, how about:

```ruby
headers = headers.dup if headers.frozen?
headers['foo'] = 'bar'
```

To avoid memory pollution if headers aren't frozen.